network/peer.go

```diff
@@ -204,31 +204,31 @@ func (p *peer) WriteMessages() {
 
 // send assumes that the stateLock is not held.
 func (p *peer) Send(msg Msg) bool {
-	p.net.stateLock.Lock()
-	defer p.net.stateLock.Unlock()
-
 	return p.send(msg)
 }
 
-// send assumes that the stateLock is held.
 func (p *peer) send(msg Msg) bool {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
 	if p.closed {
 		p.net.log.Debug("dropping message to %s due to a closed connection", p.id)
 		return false
 	}
 
 	msgBytes := msg.Bytes()
-	newPendingBytes := p.net.pendingBytes + len(msgBytes)
-	newConnPendingBytes := p.pendingBytes + len(msgBytes)
+	msgBytesLen := int64(len(msgBytes))
+	newPendingBytes := atomic.LoadInt64(&p.net.pendingBytes) + msgBytesLen
```

Comment:
I think there is a logical race introduced here. Not sure if we really care, but we should document it if we are intentionally allowing for a race.